alacritty_terminal/src/grid/storage.rs

```diff
@@ -173,25 +176,41 @@ impl<T> Storage<T> {
         }
     }
 
-    /// Truncate the invisible elements from the raw buffer
+    /// Truncate the invisible elements from the raw buffer.
     pub fn truncate(&mut self) {
         self.inner.rotate_left(self.zero);
         self.inner.truncate(self.len);
 
         self.zero = 0;
     }
 
-    /// Dynamically grow the storage buffer at runtime
+    /// Dynamically grow the storage buffer at runtime. This function tries to minimize
+    /// reallocations by using minimun step for growing internal storage. If you want to avoid this
+    /// behavior use `initialize_exact`.
     pub fn initialize(&mut self, num_rows: usize, template_row: Row<T>)
     where
         T: Clone,
     {
-        let mut new = vec![template_row; num_rows];
+        if self.len + num_rows > self.inner.len() {
+            let len = self.len;
+            self.initialize_exact(max(num_rows, MIN_INIT_SIZE), template_row);
+            // Restore proper len
+            self.len = len + num_rows;
```

Comment:
Why is this necessary? Shouldn't `initialize_exact` properly update `self.len`?

Reply:
It does the goal is not to have 2 functions with the same body, so I'm just reusing `initialize_exact`. We're not wasting anything, we just preallocate when do a reallocation in storage.rs instead of doing this kind of things in grid/mod.rs.

Reply:
I'm not following... If we allocate 1,000 new rows (`max(num_rows, MIN_INIT_SIZE)`), but `num_rows` is something less that 1,000, then we'll have left `self.len` smaller than the actual allocated range. Right?

Reply:
>I'm not following... If we allocate 1,000 new rows (max(num_rows, MIN_INIT_SIZE)), but num_rows is something less that 1,000, then we'll have left self.len smaller than the actual allocated range. Right?

Yeah, that's the whole point. `self.len` is like a `len` on std::vec::Vec, but self.inner.len() is like a `capacity` on std::vec::Vec, so the goal is to grow capacity not by 1 at a time, but by some step `MIN_INIT_SIZE`, however you should grow your len by 1 here. This wasn't the case before, so your grid.len()(from grid/mod.rs) was a self.inner.len() (from storage.rs), which is wrong.

Reply:
Ah I think I had `self.inner.len` and `self.len` backwards in my head. That makes this make more sense.